usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+    if (ret != CKR_OK) {
+        TRACE_INFO("Cannot load new MK from MK_USER, datastore probably old.\n");
+        goto done;
+    }
+
+    if (memcmp(masterkey_so, masterkey_user, 32) != 0) {
+        TRACE_ERROR("MKs from MK_SO and MK_USER don't match, datastore inconsistent.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    ret = count_objects(data_store, &num_objs, &num_old_objs);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("cannot count objects in %s.\n", data_store);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    TRACE_INFO("Found %d objects total, %d old objects.\n", num_objs, num_old_objs);
+    if (num_objs > 0 && num_old_objs == 0)
+        *new = 1;
+
+    ret = CKR_OK;
+
+done:
+
+    return CKR_OK;
+}
+
+/**
+ * Switch to new repository by deleting the old repository and renaming
+ * the backup folder to the original data store name.
+ */
+CK_RV switch_to_new_repository(const char *data_store_old, const char *data_store_new)
+{
+    char fname1[PATH_MAX];
+    CK_RV ret;
+    int rc = -1;
+
+    /* Rename original repository folder */
+    snprintf(fname1, sizeof(fname1), "%s_OLD_TO_BE_DELETED", data_store_old);
+    rc = rename(data_store_old, fname1);
+    if (rc) {
+        TRACE_ERROR("Cannot rename %s, errno=%s.\n", data_store_old, strerror(rc));
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Rename backup folder */
+    rc = rename(data_store_new, data_store_old);
+    if (rc) {
+        TRACE_ERROR("Cannot rename %s, errno=%s.\n", data_store_new, strerror(rc));
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    return ret;
+}
+
+/**
+ * Inserts the new tokversion parm in the token's slot configuration, e.g.
+ *
+ *   slot 2
+ *   {
+ *     stdll = libpkcs11_cca.so
+ *     tokversion = 3.12
+ *   }
+ */
+CK_RV update_opencryptoki_conf(CK_SLOT_ID slot_id, char *location)
+{
+    const char *parm = "tokversion = 3.12\n";
+    char dst_file[PATH_MAX], src_file[PATH_MAX], fname[PATH_MAX+20], line[80];
+    char slot[32];
+    FILE *fp_r = NULL, *fp_w = NULL;
+    CK_RV ret;
+    int rc;
+
+    /* Open current conf file for read */
+    snprintf(src_file, PATH_MAX, "%s/%s", location, "opencryptoki.conf");
+    fp_r = fopen(src_file, "r");
+    if (!fp_r) {
+        TRACE_ERROR("Cannot open %s\n", src_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Open new conf file for write */
+    snprintf(dst_file, PATH_MAX, "%s/%s", location, "opencryptoki.conf_new");
+    fp_w = fopen(dst_file, "w");
+    if (!fp_w) {
+        TRACE_ERROR("Cannot open %s\n", dst_file);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Insert/replace tokversion parm in new file */
+    snprintf(slot, sizeof(slot), "slot %ld", slot_id);
+    while (fgets(line, sizeof(line), fp_r)) {
+        if (strstr(line, slot)) {
```

Comment:
you are not handling lines that are commented out. So you might also find 'slot x' in a comment line...